sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/Internals/Profiling/ProfilingSessionEventSource.cs

```diff
@@ -83,6 +125,88 @@ protected override void OnEventCommand(EventCommandEventArgs command)
         /// Gets or sets the active profiling session ID.
         /// The value will be null if there is no active session.
         /// </summary>
-        public string? SessionId { get; private set; }
+        public string? SessionId
+        {
+            [NonEvent]
+            get => _sessionId;
+
+            [NonEvent]
+            private set
+            {
+                if (_sessionId != value)
+                {
+                    _sessionId = value;
+                    _sessionIdChanged?.Invoke(this, value);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Occurs when the session ID has changed, meaning either a session has started or stopped.
+        /// </summary>
+        /// <remarks>
+        /// If you subscribe to this event when a session has already started, then your handler
+        /// will be called immediately.
+        /// </remarks>
+        public event EventHandler<string?> SessionIdChanged
+        {
+            [NonEvent]
+            add
+            {
+                _sessionIdChanged += value;
+
+                // New subscribers are notified of an active session right away.
+                if (SessionId != null)
+                {
+                    value.Invoke(this, SessionId);
+                }
+            }
+
+            [NonEvent]
+            remove
+            {
+                _sessionIdChanged -= value;
+            }
+        }
+
+        /// <summary>
+        /// Write the attributes of a <see cref="Resource"/>.
+        /// </summary>
+        /// <param name="resource">The resource.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="resource"/> is null.</exception>
+        [NonEvent]
+        public void WriteResourceAttributes(Resource resource)
+        {
+            if (resource is null)
+            {
+                return;
+            }
+
+            if (IsEnabled(EventLevel.Informational, Keywords.ResourceAttributes))
+            {
+                IEnumerable<KeyValuePair<string, object>> attributes = resource.Attributes;
+                ResourceAttributes(attributes.Select(kvp => new KeyValuePair<string, string>(kvp.Key, SafeConvertToString(kvp.Value))));
+            }
+
+            static string SafeConvertToString(object? value)
+            {
+                const int maxLength = 128;
+                try
+                {
+                    string converted = Convert.ToString(value, InvariantCulture) ?? "";
+                    return converted.Length <= maxLength ? converted : converted.Substring(0, maxLength);
+                }
+                catch
+                {
+                    return "";
+                }
```

Comment:
what are you expecting to throw here?

Reply:
Just saw the other thread with you and Raj. Consider adding some comment here to provide context for why this is necessary.